### PR TITLE
feat(VncConsole): Introduce VncConsole component

### PR DESCRIPTION
**Pull request by mareklibra**
**What**:
Provide reusable component to access server/virtual machine via VNC protocol.

**Link to Storybook**:
Mock VNC backend is not available, so storybook is just a formal one, not showing the functionality.
Screenshots will substitute it in this case.

Link: https://mareklibra.github.io/patternfly-react/

**Additional issues**:
Depends on: ~~#160 (SerialConsole)~~

**Future work**:
If this change is accepted, higher-level `<Consoles/>` component wrapping selection of either `VncConsole` or `SerialConsole` will be implemented.

### CODE REVIEW

**Comment by mareklibra** [comment]
First use of the component will be in the `cockpit-project.org`.

**Comment by mareklibra** [comment]
Part of the pattern is the element with VNC session and action button sending Ctrl+Alt+Del to the VNC server.
Please note, the combo-box in the left upper corner is not part of the pattern - it will be included in a higher-level component later.


---
The component resizes according to backend:

**Comment by mareklibra** [comment]
The `npm test` recently fails on ES6 code under `node_modules`.
Any suggestion how to fix that?

Tests for unrelated components (like Toolbar) are failing since they import other patternfly-react components via `src/index.js` which transitively imports `src/VncConsole/index.js` and so the failing `node_modules/@novnc/novnc/core/*.js` .

Selected node_modules subdirectories need to be babelized for `jest`.

There is no such issue for `npm run build`, just for the tests.

For consuming application, it can be fixed i.e. like within https://github.com/cockpit-project/cockpit/pull/8894 (see change to [webpack.config.js](https://github.com/cockpit-project/cockpit/pull/8894/files#diff-11e9f7f953edc64ba14b0cc350ae7b9d)).

**Comment by mareklibra** [comment]
The issue is fixed similar way as for Cockpit - by adding `transformIgnorePatterns` to `jest` configuration

**Comment by mareklibra** [comment]
Remains issue with `noVNC` in missing 'MutationObserver' in `novnc/core/util/events.js`.
Will fix in the `noVNC` project.

**Comment by mareklibra** [comment]
Workaround for missing `MutationObserver` class within `jest` testing is added.
Keep trying to fix it in `noVNC`.

**Comment by mareklibra** [comment]
Added babel-loader for `node_modules/@novnc` required by storybook build

**Comment by mareklibra** [comment]
Rebased

**Comment by jeff-phillips-18** [comment]
@mareklibra with the merge of #311, we've updated where the consoles code should live. Can you please update your code to be under https://github.com/patternfly/patternfly-react/tree/master/packages/console ?

**Comment by mareklibra** [comment]
Rebased since #160 is merged.

**Comment by mareklibra** [comment]
This is ready for review.

**Comment by mareklibra** [comment]
Rebased after #332

**Comment by mareklibra** [comment]
May I ask for review, please?

File: packages/console/src/VncConsole/VncActions.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Button, noop } from 'patternfly-react';
+
+const VncActions = ({ textSendShortcut, textCtrlAltDel, onCtrlAltDel }) => (
+  <span className="console-actions-pf">
+    {textSendShortcut}
+
+    <Button
+      bsClass="btn btn-default console-actions-buttons-pf"
```
Comment (jeff-phillips-18) [inline packages/console/src/VncConsole/VncActions.js]: no need for btn or btn-default

**Comment by jeff-phillips-18** [comment]
Please use `yarn commit` to create the commit message. With the merge of #332 the commit message will need specific info.

File: packages/console/src/VncConsole/VncActions.js (hunk 1 above)
Comment (mareklibra) [reply packages/console/src/VncConsole/VncActions.js]: done

**Comment by mareklibra** [comment]
Commit messages updated

**Comment by mareklibra** [comment]
Rebased

File: package.json [hunk 2]
```diff
@@ -72,6 +72,7 @@
     "lerna-semantic-release": "^9.1.0",
     "lint-staged": "^7.0.4",
     "lodash": "^4.17.0",
+    "mutation-observer": "^1.0.3",
```
Comment (priley86) [inline package.json]: this can be moved to `packages/console/package.json` dependencies now i believe.

Comment (priley86) [reply package.json]: sorry - this is only used in tests. Nevermind ;)

**Comment by priley86** [comment]
do you have plans to update patternfly-react dependency in react-console to latest? It is currently targeting 1.19.1.

```
  "dependencies": {
    "@novnc/novnc": "^1.0.0",
    "classnames": "^2.2.5",
    "mutation-observer": "^1.0.3",
    "patternfly": "^3.45.0",
    "patternfly-react": "^1.19.1",
```

File: package.json (hunk 2 above)
Comment (mareklibra) [reply package.json]: yes, jest tests are configured once per project

**Comment by mareklibra** [comment]
`patternfly-react` dependency version updated to `^2.3.3`

**Comment by priley86** [comment]
This is a pretty significant test coverage reduction (-18%) but I am OK to merge the PR if this is needed quickly downstream. If you have any chance to add more tests @mareklibra, it is always appreciated.

**Comment by mareklibra** [comment]
Rebased (conflict in `yarn.lock`).
Tests added for `<VncActions>`.

@priley86 , I would love to add more thorough test for the `<VncConsole>` but this would probably require writing mock VNC server which is pretty complex task for this moment.

**Comment by beanh66** [comment]
@mareklibra Can we use the [PF Toolbar](http://www.patternfly.org/pattern-library/forms-and-controls/toolbar/) above the console area, instead of having a floating button? Also wondering if we can change the interaction to just be a dropdown called "Send Key" with the possible shortcuts available for selection. I am showing the hover state in the mockup below: 
ctrl alt del

**Comment by serenamarie125** [comment]
+1 @beanh66 thanks!

**Comment by mareklibra** [comment]
@beanh66 , sure, makes sense.

**Comment by mareklibra** [comment]
Working on that but hit an issue with Cockpit since the patternfly-react's Toolbar uses features of React 16.2 unavailable in react-lite which is used by Cockpit (conforms scope of React 15.x).
Cockpit can be either upgraded to full React or the Toolbar can not be used there.

**Comment by mareklibra** [comment]
After discussion in mailing list, I'm trying to upgrade `react-lite` to `React 16.4` in Cockpit.
I believe, this change will be beneficial for the project even in other cases.

https://github.com/cockpit-project/cockpit/pull/9263

File: packages/console/src/VncConsole/VncActions.test.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  expect(view).toMatchSnapshot();
+});
+
+test('VncActions renders correctly custom text', () => {
+  const view = shallow(
+    <VncActions
+      textSendShortcut="My Send Shortcut description"
+      textCtrlAltDel="foobar"
+      onCtrlAltDel={jest.fn()}
+    />
```
Comment (jeff-phillips-18) [inline packages/console/src/VncConsole/VncActions.test.js]: We should simulate a click and verify the onCtrlAltDel function gets called

File: packages/console/src/VncConsole/VncConsole.test.js [hunk 4]
```diff
@@ -4,6 +4,6 @@ import { shallow } from 'enzyme';
 import VncConsole from './VncConsole';
 
 test('placeholder render test', () => {
-  const view = shallow(<VncConsole />);
+  const view = shallow(<VncConsole host="my.unknown.host" />);
   expect(view).toMatchSnapshot();
```
Comment (jeff-phillips-18) [inline packages/console/src/VncConsole/VncConsole.test.js]: Can we add tests to modify some of the props and verify the expected results? Simulate the connecting, disconnected, and connected states, etc.

File: packages/console/src/VncConsole/VncConsole.js [hunk 5]
```diff
@@ -1,11 +1,180 @@
...
-const VncConsole = () => <div>VncConsole</div>;
+import { noop } from 'patternfly-react';
 
-VncConsole.propTypes = propTypes;
-VncConsole.defaultProps = defaultProps;
+import VncActions from './VncActions';
+
+const CONNECTING = 'connecting';
+const CONNECTED = 'connected';
+const DISCONNECTED = 'disconnected';
```
Comment (ohadlevy) [inline packages/console/src/VncConsole/VncConsole.js]: afair there are more states? e.g. https://github.com/theforeman/foreman/blob/develop/app/assets/javascripts/noVNC.js#L14

File: packages/console/src/VncConsole/VncConsole.js [hunk 6]
```diff
@@ -1,11 +1,180 @@
...
+  };
+
+  onDisconnected = e => {
+    this.setState({ status: DISCONNECTED });
+    this.props.onDisconnected(e);
+  };
+
+  onSecurityFailure = e => {
+    this.setState({ status: DISCONNECTED });
+    this.props.onSecurityFailure(e);
```
Comment (ohadlevy) [inline packages/console/src/VncConsole/VncConsole.js]: maybe I'm missing it, but do you capture the failure text somewhere? e.g. like at https://github.com/theforeman/foreman/blob/develop/app/assets/javascripts/noVNC.js#L27 ?

**Comment by mareklibra** [comment]
@dabeng, thanks for reporting that. Since it's about different component, I filed new issue for that.

File: packages/console/src/VncConsole/VncConsole.js [hunk 7]
```diff
@@ -1,11 +1,180 @@
...
+      vncLogging,
+      onInitFailed
+    } = this.props;
+
+    NovncLog.init_logging(vncLogging);
+    try {
+      const protocol = encrypt ? 'wss' : 'ws';
+      const url = `${protocol}://${host}:${port}/${path}`;
+
+      const options = {
```
[the remark this answers is not in the document]
Comment (mareklibra) [reply packages/console/src/VncConsole/VncConsole.js]: Not sure if these are still valid options: https://github.com/novnc/noVNC/blob/master/docs/API.md#rfb-1

When looking at https://github.com/novnc/noVNC/blob/master/core/rfb.js, they are handled differently in recent version.

File: packages/console/src/VncConsole/VncConsole.js (hunk 5 above)
Comment (mareklibra) [reply packages/console/src/VncConsole/VncConsole.js]: These states reflect events received from the RFB (https://github.com/novnc/noVNC/blob/master/docs/API.md#events).

This seems to be changed as well.
For example, in https://github.com/novnc/noVNC/blob/0342e4f4891a0c29f10b25a53258bf40890bc9fd/core/rfb.js#L675 , the failure is wrapped behind the `disconnected` state.

File: packages/console/src/VncConsole/VncConsole.js (hunk 6 above)
Comment (mareklibra) [reply packages/console/src/VncConsole/VncConsole.js]: Not recently, this goes up.

**Comment by mareklibra** [comment]
Rebased

File: packages/console/src/VncConsole/VncConsole.js (hunk 6 above)
Comment (mareklibra) [reply packages/console/src/VncConsole/VncConsole.js]: Any idea how to handle the textual information within the component? I18n needs to be considered...

**Comment by mareklibra** [comment]
Rebased after #394

**Comment by mareklibra** [comment]
Removing `yarn.lock` from this PR did the trick.
The series of rebases broke it.

**Comment by mareklibra** [comment]
Test for `<VncActions />` added

File: package.json [hunk 8]
```diff
@@ -158,5 +161,8 @@
       "eslint --fix",
       "git add"
     ]
+  },
+  "dependencies": {
+    "mutation-observer": "^1.0.3"
```
Comment (jeff-phillips-18) [inline package.json]: Should this be a devDependency?

File: packages/console/less/console.less [hunk 9]
```diff
@@ -3,13 +3,3 @@
...
-
-.console-actions-pf {
-  float: right;
-  position: relative;
-  top: -2em;
-}
-
-.console-actions-buttons-pf {
-  margin-left: 5px;
-}
```
Comment (jeff-phillips-18) [inline packages/console/less/console.less]: These classes are used by `SerialConsole`

File: packages/console/src/VncConsole/VncConsole.js [hunk 10]
```diff
@@ -1,11 +1,180 @@
...
+        status = <div className="vnc-console-connecting">{textConnecting}</div>;
+    }
+
+    if (!this.novncStaticComponent) {
+      // create just once
+      this.novncStaticComponent = <div ref={this.setNovncElem} />;
+    }
+
+    return (
+      <div className="vnc-console">
```
Comment (jeff-phillips-18) [inline packages/console/src/VncConsole/VncConsole.js]: We should allow a className to be added by callers here.

Comment (mareklibra) [reply packages/console/src/VncConsole/VncConsole.js]: ok

File: packages/console/less/console.less (hunk 9 above)
Comment (mareklibra) [reply packages/console/less/console.less]: In the meantime, I moved them to `serial-console` styling

**Comment by mareklibra** [comment]
@jeff-phillips-18 , done

**Comment by serenamarie125** [comment]
@mareklibra as soon as you update the screenshots, please tag @beanh66 and I so we can provide the UX review.  I've added the label for awareness

**Comment by mareklibra** [comment]
Added fix for overflowing `DropdownButton` when used under `Toolbar.RightContent`

**Comment by mareklibra** [comment]
@beanh66 , recent screenshots:

Narrow display, collapsed:

01_narrow_collapsed
---
Narrow display, expanded:
02_narrow_expanded

---
Wide display, expanded:
03_wide_expanded

---

Please note, the blurred `Console Type` dropdown is not part of this pattern. Will be added later.

**Comment by mareklibra** [comment]
How it looks like in Cockpit with an installer running:

04_cockpit_installer

**Comment by beanh66** [comment]
@mareklibra Could we put the dropdown in a toolbar or add the line separator to mimic a toolbar? Per our previous discussion, I thought with the react upgrade this would now be possible. Other question is around the button placement...can we right align the button with the edge of the black console area? 

FYI @lizsurette

**Comment by mareklibra** [comment]
> Could we put the dropdown in a toolbar

@beanh66 , actually the toolbar was already there. To render the separator (as part of the `Toolbar` pattern), I've moved the console under `Toolbar.Results` component.

The line separator recently "overlaps" content, as implemented within the `Toolbar.Results`. Please let me know, if this is ok. I would change it otherwise - either by overriding the `Toolbar` pattern styling or by avoiding the `Toolbar.Results` and rendering the separator differently.

> Can we right align the button with the edge of the black console area?

Done.

> Can we add the other shortcut options in this menu? I think Michal mentioned we would have at least one other option in there.

@michalskrivanek , ideas?
I.e. Alt-F4 ? Ctrl-W ?

**Comment by mareklibra** [comment]
Pattern recently:
01_pattern

---

As used within Cockpit:

02_cockpit

**Comment by michalskrivanek** [comment]
> @michalskrivanek , ideas?
> I.e. Alt-F4 ? Ctrl-W ?

take a look at virt-viewer;-)

**Comment by mareklibra** [comment]
>> I.e. Alt-F4 ? Ctrl-W ?

> take a look at virt-viewer;-)

So what to include **in the pattern**? All the shurtcuts implemented within virt-viewer?
It would mean: Ctrl+Alt+Backspace , Ctrl+Alt+Del , Ctrl+Alt+F1 .. F12 , Printscreen

In addition, Alt-F4 is handed by browser recently, causing it's window to be closed.

Or provide an extension mechanism to add additional ones?

**Comment by serenamarie125** [comment]
@mareklibra might it be possible to allow for the following by default Ctrl+Alt+Backspace , Ctrl+Alt+Del  & Printscreen (but allow them to be not shown)?  You could also allow for a mechanism to add additional commands, but that could be a future contribution as well.

File: packages/console/src/VncConsole/VncActions.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+  textCtrlAltDel: PropTypes.string,
+  textSendShortcut: PropTypes.string
+};
+
+VncActions.defaultProps = {
+  onCtrlAltDel: noop,
+
+  textCtrlAltDel: 'Ctrl+Alt+Del',
+  textSendShortcut: 'Send key'
```
Comment (serenamarie125) [inline packages/console/src/VncConsole/VncActions.js]: please use initial caps ... "Send Key"

File: packages/console/src/VncConsole/__snapshots__/VncActions.test.js.snap [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+`;
+
+exports[`VncActions renders correctly html 1`] = `"<div class=\\"dropdown btn-group btn-group-default\\"><button id=\\"console-send-shortcut\\" role=\\"button\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" type=\\"button\\" class=\\"dropdown-toggle btn btn-default\\">My Send Shortcut description <span class=\\"caret\\"></span></button><ul role=\\"menu\\" class=\\"dropdown-menu\\" aria-labelledby=\\"console-send-shortcut\\"><li role=\\"presentation\\" class=\\"\\"><a role=\\"menuitem\\" tabindex=\\"-1\\" href=\\"#\\">foobar</a></li></ul></div>"`;
+
+exports[`placeholder render test 1`] = `
+<DropdownButton
+  bsStyle="default"
+  id="console-send-shortcut"
+  onClick={[Function]}
+  title="Send key"
```
Comment (serenamarie125) [inline packages/console/src/VncConsole/__snapshots__/VncActions.test.js.snap]: please use initial caps ... "Send Key"

File: packages/console/src/VncConsole/VncActions.js (hunk 11 above)
Comment (mareklibra) [reply packages/console/src/VncConsole/VncActions.js]: done

File: packages/console/src/VncConsole/__snapshots__/VncActions.test.js.snap (hunk 12 above)
Comment (mareklibra) [reply packages/console/src/VncConsole/__snapshots__/VncActions.test.js.snap]: done

**Comment by mareklibra** [comment]
@serenamarie125 , this makes makes sense.
I can add mechanism to append custom combinations with handful predefined ones.

If it is ok, I would postpone that work to a follow-up to avoid further prolonging merge of this.

**Comment by mareklibra** [comment]
Rebased.
Build is recently failing on unrelated issues
- ApplicationLauncherWrapper
- styling

**Comment by serenamarie125** [comment]
> If it is ok, I would postpone that work to a follow-up to avoid further prolonging merge of this.

+1 sounds good, thank you!  I've created https://github.com/patternfly/patternfly-react/issues/422 to track the enhancement.

File: package.json [hunk 13]
```diff
@@ -167,6 +168,9 @@
     "roots": [
       "<rootDir>/packages",
       "<rootDir>/scripts"
+    ],
+    "transformIgnorePatterns": [
+      "node_modules/(?!@novnc)"
```
Comment (dmiller9911) [inline package.json]: this needs to be added to the one on line 165.  Looks like a rebase issue.
```
 "transformIgnorePatterns": [
       "node_modules/(?!@patternfly)",
       "node_modules/(?!@novnc)"
  ],
```

Comment (dmiller9911) [reply package.json]: That should fix the build

Comment (mareklibra) [reply package.json]: right, a rebase issue.
Thanks for spotting this.

**Comment by mareklibra** [comment]
Rebased and fixed the former rebase issue with package.json

**Comment by mareklibra** [comment]
May I ask for (final) review? @dmiller9911 , @jeff-phillips-18

**Comment by jeff-phillips-18** [comment]
Going to merge this though we dropped coverage. Hopefully we can address this soon.

**Pull request merged**